File: Data_Structure/DS_Using_Python/Insertion_In_Sorted_Doubly_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		self.prev = None

class LinkedList:
	def __init__(self):
		self.head = None
			
	def push(self, data):
		newNode = Node(data)
		newNode.next = self.head
		self.head = newNode
		
	def printList(self, ptr):
		ptr = self.head
		if ptr is None:
			print('List has no element')
			return
		else:
			temp = ptr
			while(temp):
				print(temp.data, end=' ')
				temp = temp.next
				
	def sortedNodeLinkedList(self, item):
				newNode = Node(item)
				if self.head is None:
					newNode.data = item
					newNode.next = self.head
					newNode.prev = None
					self.head.prev = newNode
					self.head = newNode
				elif self.head.data >= item:
					newNode.data = item
					newNode.next = self.head
					newNode.next.prev = newNode
					self.head = newNode
				else:
					temp = self.head
					while((temp.next is not None) and (temp.next.data < item)):
						temp = temp.next
					newNode.data = item
					newNode.next = temp.next
					if temp.next is not None:
						newNode.next.prev = newNode
					temp.next = newNode
					newNode.prev = temp
				
llist = LinkedList()

llist.push(92)
llist.push(70)
llist.push(36)
llist.push(28)
llist.push(12)

print("Linked List is")
llist.printList(llist.head)

print("\n\nNew Linked List")
llist.sortedNodeLinkedList(3)
llist.sortedNodeLinkedList(156);
llist.sortedNodeLinkedList(23);
llist.printList(llist.head)

File: Data_Structure/LinkedList/Insertion_In_Sorted_Doubly_LinkedList.c
//Insertion of a node to a sorted doubly linked list

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
    struct node * prev;
};

struct node *head = NULL;

void linkedListTraversal(struct node* ptr)
{
    ptr = head;
    while(ptr != NULL)
    {
        printf("%d ", ptr -> data);
        ptr = ptr -> next;
    }
}

void createList(int data)
{
    struct node *ptr = (struct node*)malloc(sizeof(struct node));
    ptr -> data = data;
    ptr -> next = head;	
    head = ptr;
}

void sortedNodeLinkedList(struct node** p, int item)
{
    struct node* ptr = (struct node*)malloc(sizeof(struct node));
    if((*p) == NULL || (*p) -> data >= item)
    {
        ptr -> data = item;
        ptr -> next = *p;
        ptr -> prev = NULL;
        (*p) -> prev = ptr;
        (*p) = ptr;
        return;
    }
    printf("%d ", (*p) -> data);
    struct node* q = *p;
    while((q -> next != NULL) && (q -> next -> data < item))
    {    
        q = q -> next;   
    }
    ptr -> data = item;
    if(q -> next != NULL)
    {
       q -> next -> prev = ptr;
    }
    ptr -> next = q -> next;
    ptr -> prev = q;
    q -> next = ptr;
}


int main()
{
   
    createList(92);
    createList(70);
    createList(36);
    createList(28);
    createList(12);
    printf("Linked List is\n");
    linkedListTraversal(head);
    
    sortedNodeLinkedList(&head, 1);
    sortedNodeLinkedList(&head, 356);
    sortedNodeLinkedList(&head, 23);
    printf("\n\nNew Linked List is\n");
    linkedListTraversal(head);
    return 0;
}